pcp/PCPDynamicColumn.c

```diff
@@ -0,0 +1,330 @@
+      free(note);
+   } else { /* overwrite closing brace */
+      *p = '\0';
+   }
+}
+
+static PCPDynamicColumn* PCPDynamicColumn_new(PCPDynamicColumns* columns, const char* name) {
+   PCPDynamicColumn* column = xCalloc(1, sizeof(*column));
+   String_safeStrncpy(column->super.name, name, sizeof(column->super.name));
+   Hashtable_put(columns->table, ++columns->count, column);
+   return column;
+}
+
+static void PCPDynamicColumn_parseFile(PCPDynamicColumns* columns, const char* path) {
+   FILE* file = fopen(path, "r");
+   if (!file)
+      return;
+
+   PCPDynamicColumn* column = NULL;
+   unsigned int lineno = 0;
+   for (;;) {
+      char* line = String_readLine(file);
+      if (!line)
+         break;
+      lineno++;
+
+      /* cleanup whitespace, skip comment lines */
+      char* trimmed = String_trim(line);
+      free(line);
+      if (trimmed[0] == '#' || trimmed[0] == '\0') {
+         free(trimmed);
+         continue;
+      }
+
+      size_t n;
+      char** config = String_split(trimmed, '=', &n);
+      free(trimmed);
+      if (config == NULL)
+         continue;
+
+      char* key = String_trim(config[0]);
+      char* value = n > 1 ? String_trim(config[1]) : NULL;
+      if (key[0] == '[') {  /* new section heading - i.e. new column */
+         PCPDynamicColumn_validateColumnName(key + 1, path, lineno);
+         column = PCPDynamicColumn_new(columns, key + 1);
+      } else if (value && String_eq(key, "caption")) {
+         free_and_xStrdup(&column->super.caption, value);
+      } else if (value && String_eq(key, "description")) {
+         free_and_xStrdup(&column->super.description, value);
+      } else if (value && String_eq(key, "width")) {
+         column->super.width = strtoul(value, NULL, 10);
+      } else if (value && String_eq(key, "maximum")) {
+         column->super.maximum = strtod(value, NULL);
+      } else if (value) {
+         PCPDynamicColumn_parseMetric(columns, column, path, lineno, key, value);
+      }
+      String_freeArray(config);
+      free(value);
+      free(key);
+   }
+   fclose(file);
+}
+
+static void PCPDynamicColumn_scanDir(PCPDynamicColumns* columns, char* path) {
+   DIR* dir = opendir(path);
+   if (!dir)
+      return;
+
+   struct dirent* dirent;
+   while ((dirent = readdir(dir)) != NULL) {
+      if (dirent->d_name[0] == '.')
+         continue;
+
+      char* file = String_cat(path, dirent->d_name);
+      PCPDynamicColumn_parseFile(columns, file);
+      free(file);
+   }
+   closedir(dir);
+}
+
+void PCPDynamicColumns_init(PCPDynamicColumns* columns) {
+   const char* sysconf = pmGetConfig("PCP_SYSCONF_DIR");
+   const char* xdgConfigHome = getenv("XDG_CONFIG_HOME");
+   const char* home = getenv("HOME");
+   char* path;
+
+   columns->table = Hashtable_new(0, true);
+
+   /* search in the users home directory first of all */
+   if (xdgConfigHome) {
+      path = String_cat(xdgConfigHome, "/htop/columns/");
+   } else {
+      if (!home)
+         home = "";
```

Comment:
This effectively tries to read from the filesystem root, which is unlikely what's intended …

Reply:
This behaviour is mirroring the way Settings_new() works over in Settings.c ... I agree its a bit odd though.  Will tweak it.